src/ngx_http_lua_kong_log.c

```diff
@@ -14,45 +14,150 @@
  * limitations under the License.
  */
 
-
 #include "ngx_http_lua_kong_common.h"
 
+#if (NGX_WIN32)
+/*
+** The dynamic log level feature is only supported on Linux and macOS,
+** because we only call the function
+** void ngx_http_lua_kong_log_set_ready_for_dynamic_log_level();
+** in ngx_worker_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+**
+** Excluding the technical view,
+** the Kong Gateway does not supoort the Windows OS.
+** 
+*/
+#error "The dynamic log level feature is only supported on Linux and macOS"
+#endif
+
+#if (UINT32_MAX < NGX_LOG_DEBUG_FIRST)  || \
+    (UINT32_MAX < NGX_LOG_STDERR)       || \
+    (UINT32_MAX < NGX_LOG_EMERG)        || \
+    (UINT32_MAX < NGX_LOG_ALERT)        || \
+    (UINT32_MAX < NGX_LOG_CRIT)         || \
+    (UINT32_MAX < NGX_LOG_ERR)          || \
+    (UINT32_MAX < NGX_LOG_WARN)         || \
+    (UINT32_MAX < NGX_LOG_NOTICE)       || \
+    (UINT32_MAX < NGX_LOG_INFO)         || \
+    (UINT32_MAX < NGX_LOG_DEBUG)
+#error "Some of the log levels are too large for uint32_t"
+#endif
+
+#if (NGX_OK > INT32_MAX)    ||     \
+    (NGX_OK < INT32_MIN)    ||     \
+    (NGX_ERROR > INT32_MAX) ||     \
+    (NGX_ERROR < INT32_MIN)
+#error "The int32_t type is not suitable for NGX_OK and NGX_ERROR"
+#endif
+
+static ngx_int_t g_ready_for_dynamic_log_level = 0;
+
+void
+ngx_http_lua_kong_log_set_ready_for_dynamic_log_level(ngx_int_t is_ready)
+{
+    /*
+    ** This function will be called when the following functions
+    ** are about to start event loop:
+    **   * ngx_worker_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_single_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_manager_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_loader_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_privileged_agent_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **
+    ** This function will be called immediately when the event loop has just been stopped:
+    **   * ngx_worker_process_exit in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_master_process_exit in the file src/os/unix/ngx_process_cycle.c.
+    **
+    ** Once the event loop is started,
+    ** we can confirm the data structure and other resources are ready for
+    ** the dynamic log level feature.
+    **
+    ** Please see the function ngx_worker_process_cycle in the file
+    ** src/os/unix/ngx_process_cycle.c.
+    */
+    g_ready_for_dynamic_log_level = is_ready;
+}
 
-int
-ngx_http_lua_kong_ffi_set_log_level(ngx_http_request_t *r, int level)
+int32_t
+ngx_http_lua_kong_ffi_set_dynamic_log_level(uint32_t log_level, uint32_t timeout)
```

Comment:
I think the pararm could be simple `int level`, could you refer to lua-nginx-module `ngx_http_lua_ffi_errlog_set_filter_level` in `ngx_http_lua_log.c`?